### ISSUE TITLE
Put integer types into order

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1450]
Currently, igraph uses the following types for integers:

 - `igraph_real_t`, i.e. `double`: #556 and #1255 touch on this.
 - `igraph_integer_t` which is defined as `int`. In practice, this is currently only 32-bit even on 64-bit platforms.
 - `long`
 - internally, occasionally `int`s are used too

We should standardize the use of integer types. Here are the reasons:

 - The built-in types of C (`int`, `long`) have generally unpredictable sizes. For example, on 64-bit operating systems, `long` is 64-bit on macOS and x86 Linux, but it is 32-bit on Windows. _We should have some consistency across platforms._
 - The use of integer types is commonly tied to indexing into arrays. The C standard library provides `size_t` and `ptrdiff_t` for these. igraph tends to use `long` internally, which I assume was born out of the assumption that this would match the address size of the platform, but this is not the case in general (see `long` being 32-bit on Windows). Those internal `long`s are often cast to `igraph_integer_t`, which is `int`, and therefore the cast results in truncation.  _igraph should have an analogue of `ptrdiff_t` which matches the address size of the system._
 - _There should be guidelines on what integer type to use in what situation._ Right now, someone implementing new functions would be quite lost about whether to use `igraph_integer_t`, `int`, `long` or `size_t`. There should be some clarity about what should be used in a given situation and why.
 - We need to have some certainly about the sizes of integers before we can effectively tackle issues with integer overflow. See #51, #1446, and especially #1435.
 - _We should have some clarity about igraph's limits._ What is the largest supported edge count and vertex count? There should be checks to prevent exceeding these limits. There should be `IGRAPH_MAX_VCOUNT` and `IGRAPH_MAX_ECOUNT` constants provided. See #1435, a real-world use case where a large graph is successfully created, but its edge count cannot be retrieved with `igraph_ecount` as it does not fit into `igraph_integer_t`

----

**I would prefer the following solution for its simplicity:**

 - Let `igraph_integer_t` be a signed 32-bit integer on 32-bit platforms and 64-bit on 64-bit platforms (i.e. not simply use `int` or `long` but target a specific size that matches the architecture).
 - Use only `igraph_integer_t` in all situations, including for indexing into arrays. Effectively, `igraph_integer_t` will be the analogue of `ptrdiff_t` (signed value) and we don't bother with an unsigned analogue of `size_t`.
 - Therefore, all array types will have a maximum size of `2^31 - 1` or `2^63 - 1`. The maximum graph sizes follow from this.

----

**Largest possible graph sizes:**

Let's think about the largest reasonable graph sizes that can be supported. Let's denote the vertex and edge counts by `V` and `E`. Currently, an `igraph_t` contains 4 arrays of size `E` and 2 arrays of size `V+1`. Let's denote the address size of the platform by `w` (w=32 or w=64). Then,

 - Max vertex count cannot be larger than `2^(w-1) - 2` (since `igraph_t` has arrays of size `V+1`).
 - Max edge count cannot be larger than `2^(w-1) - 1`, but since graphs are often created from a vector which contains the head and tail of each edge, effectively we would be typically be restricted to `2^(w-2) - 1`.
 - Given that typically the maximum addressable memory is `2^w`, to fit all four arrays of size E in memory, assuming an element size of `w/8`, we must have `E < 2^(w+1) / w`, which on 32-bit is 2^28 = 2^(32-4) = 268,435,456. A graph of this size will still exceed the addressable memory in practice.

Therefore, considering that in practical cases usually V < E, we could set both the maximum allowed vertex count and edge count to be `2^(w-4)`. This limit is relaxed enough that graphs of this size still cannot be created on a 32-bit platform (i.e. graph sizes will be limited by addressable memory and not arbitrary limits imposed by igraph), but it is tight enough that simply checking that `V < V_max` will automatically protect against some cases of overflow (such as the one in `igraph_star` #1446). 

For better overflow protection, on 64-bit we might even reduce the limit arbitrarily to `2^(w-7)`, which would still allow graphs that require hundreds of petabytes to store (i.e. the real practical constraint will still be the physical memory that is available).

**Comment by iosonofabio** [comment]
I love this issue! I have been so confused about this forever.

I'm in favour of not bothering with the factor 2 of `size_t`. However I'm tempted to say let's just have `igraph_integer_t` be a 64 bit signed on all platforms, whether 32 or 64. A little slower if on a 32bit platform, but larger graphs will fit.

**Comment by szhorvat** [comment]
> However I'm tempted to say let's just have `igraph_integer_t` be a 64 bit signed on all platforms, whether 32 or 64. A little slower if on a 32bit platform, but larger graphs will fit.

Having larger integers will in fact limit the size of graphs even more (maximum graph size will be smaller, not larger). In short, this is because the practical limiting factor to graph sizes on 32-bit is addressable memory. Currently, `igraph_t` uses vectors of `double`s, but let's assume we will transition to integers. If we use 32-bit integers (4 bytes), and the max addressable memory is 2^32 = 4 GB, then at most 2^32 / 4 = 2^30 integers will fit. If we use 64-bit integers, then only 2^29 of them will fit. So, in practice, smaller integer size ==> bigger limit on graph sizes.

(In practice, the max addressable memory seems to be 2^32 = 4 GB on ARM, at least on my Raspberry Pi, and 2^31 = 2 GB on x86, unless the uncommon and typically server-only [PAE](https://en.wikipedia.org/wiki/Physical_Address_Extension) feature is enabled. But servers these days are 64-bit—in practice when igraph is used on 32-bit x86, the 2 GB limit is likely to apply.)

**Comment by szhorvat** [comment]
@gaborcsardi This is why [I asked you on the forum](https://igraph.discourse.group/t/why-is-igraph-bool-t-an-int-and-not-a-char/126/8?u=szhorvat) if the R interface has any constraints on moving `igraph_integer_t` to 64-bit. It seems [people are already hitting the limits](https://github.com/igraph/igraph/issues/1435#issuecomment-664058584). 

Do I understand it correctly that R/igraph maps R arrays to igraph arrays directly (i.e. does not copy them, just uses the memory as-is), and R's integer type is still 32-bit?

It seems that this would be the biggest obstacle to modernizing integer handling in igraph.

**Comment by szhorvat** [comment]
@gaborcsardi Assuming that my understanding of the problem is correct (i.e. igraph must match R's own type in order to use arrays directly), there's an alternative solution: basically do everything I suggested above, and standardize on a single `igraph_integer_t` type, *but make that type configurable*. 

This way we would not need to compromise: people can work with big graphs from C, the R interface won't need time-consuming  and performance-decreasing changes.

I wasn't here at the beginning, but it seems to me that igraph was always designed to be accessed from high level languages. It would make perfect sense to allow matching each language's native integer type, and avoid having to copy arrays as much as possible. Currently, in the Mathematica interface, I do have to copy integer arrays, but not `double` arrays. It would be great if I could avoid copying integer arrays too. @ntamas How does the Python interface handle this?

I see two potential disadvantages to making `igraph_integer_t` configurable:

 - This sort of flexibility can be a burden. What if the code works with one choice for `igraph_integer_t`, but not the other? I think this is manageable: basically we only need to deal with 32- and 64-bit signed integers, that's all. We don't need to support anything else in practice. 

 - Distro maintainers might scream bloody murder if they can't link every high level interface to the same C core library. Personally I'm in favour of static linking anyway ... it gives more flexibility for the high-level interfaces about which specific version of the C core to use. I think there should be a reasonable solution that pacifies distro maintainers. Have two supported configuration (32- and 64-bit types) with two different names for the binary library?

Any thoughts?

**Comment by ntamas** [comment]
> Do I understand it correctly that R/igraph maps R arrays to igraph arrays directly (i.e. does not copy them, just uses the memory as-is), and R's integer type is still 32-bit?

I think that the R interface [copies the vector as well](https://github.com/igraph/rigraph/blob/dev/tools/stimulus/rinterface.c.in#L2381), at least in the generic case. It would probably need to do so anyway because `igraph_integer_t` is now an `int`, and I think R uses doubles internally (there is no distinction between integer vectors and non-integer vectors).

The harder part is probably the conversion between an igraph graph on R's side and an igraph graph on the C side. The R representation of an igraph graph contains all the vectors that the data structure on the C side requires, and these vectors are indeed just copied as-is, so the vectors in igraph's C side representation in `type_indexededgelist.c` need to be compatible with R. See [this part](https://github.com/igraph/rigraph/blob/dev/tools/stimulus/rinterface.c.in#L2640) in the source code of the R interface. So, every time an igraph function is invoked from the R interface, we copy six vectors of size O(n+m) if the graph has _n_ vertices and _m_ edges. I don't know why it is done this way in the R interface; in Python, the `Graph` object simply contains an `igraph_t` and everything is managed on the C side (at the expense of slower serialization if the `Graph` object has to be pickled or stored by other means). But this whole issue affects the treatment of `igraph_real_t`, not `igraph_integer_t`.

> I wasn't here at the beginning, but it seems to me that igraph was always designed to be accessed from high level languages. 

igraph was initially an R-only library with a C core. Eventually the C core was separated into a library on its own, and that was when I joined the project and started developing a Python interface for it (because Gábor was using R for research and I was using mostly Python, but we both worked with networks). Back in the very early days we even had 1-based indexing in the C core (because that's what R uses), and that is also the reason why `igraph_integer_t` was also a `double` for a long time instead of an `int` (for sake of compatibility with R). Anyway, that's history, at some point we realized that we cannot design a low-level C library around the needs and peculiarities of one particular high-level interface (R) so we switched `igraph_integer_t` to an `int`.

>  @ntamas How does the Python interface handle this?

I think I answered it already above: the `Graph` object on the Python side wraps an `igraph_t` object that is owned by the `Graph` instance, so when the GC kicks in in Python and calls the `__del__()` method of the `Graph`, we call `igraph_destroy()` on the wrapped graph. For the Python interface, the internals of `igraph_t` are a black box and we never mess around with it directly.

As for vectors, we convert back-and-forth between Python's numeric types and the underlying C types as needed.

---

As for making `igraph_integer_t` configurable: given that the R interface copies integer vectors and converts them to doubles anyway, I see no reason why we should make `igraph_integer_t` configurable, considering that we gain nothing of value but we make it a nightmare for distro maintainers if they have to deal with "32-bit igraph" and "64-bit igraph" (just like some libraries had Unicode versions and non-Unicode versions back in the dark ages). We should simply say that `igraph_integer_t` matches the size of `size_t` and we are probably okay.

And yes, we do need `IGRAPH_MAX_EDGE_COUNT` and `IGRAPH_MAX_VERTEX_COUNT` constants. We should probably also provide `IGRAPH_MAX_INTEGER` and `IGRAPH_MIN_INTEGER` to let the user query the limits of `igraph_integer_t` directly at compilation time.

**Comment by szhorvat** [comment]
[@gaborcsardi said](https://igraph.discourse.group/t/why-is-igraph-bool-t-an-int-and-not-a-char/126/8?u=szhorvat) that it might be a lot of work to adapt the R interface to a 64-bit `igraph_integer_t`. I suggested making it configurable as a compromise, to avoid having to do this work.

I assumed that there really are some requirements to match types between R and igraph. [I was told that `igraph_boolean_t` is an `int` (and not a 1-byte type) for the same reason.](https://igraph.discourse.group/t/why-is-igraph-bool-t-an-int-and-not-a-char/126/2?u=szhorvat)

I think we really need some feedback from @gaborcsardi before we can proceed.

**Comment by szhorvat** [comment]
There are several intertwined issue here, so let us break it down.

**Explicit limits on graph and array sizes**

I suggested to define and enforce explicit limits on graph and array sizes. This will help catch certain kinds of errors.

I think everyone agrees that doing this is good, we just need to settle on the details. (If not, shout.)

**Move igraph_integer_t to 64-bit**

The benefit: larger graph sizes and staying competitive with other libraries. It seemed to me that most people would support this, but there's a problem:

Apparently, this is an issue for the R interface. Perhaps it's doable, but @gaborcsardi said that it is a lot of work, and he already does not seem to have a lot of time to work on igraph. The R interface is by far the most popular way to use igraph, and a major contributor to igraph's success. IMO its health is very important to us. I do not want to put extra burdens on it. This is why I suggested making the size of `igraph_integer_t` configurable.

**Making `igraph_integer_t` configurable**

For us, I think this would not be a big maintenance burden.

I wrote that it may be an issue for Linux distros based on the false assumption that Linux distros always want to enforce linking _dynamically_ to one single copy of a library. However, this is not the case. I just verified that the R/igraph included in Ubuntu 20.04 links statically to C/igraph.

> but we make it a nightmare for distro maintainers if they have to deal with "32-bit igraph" and "64-bit igraph"

Absolutely, I should not even have mentioned this (it was actually a misunderstanding). We should definitely not make distro maintainers' work harder.

**A concrete proposal**

 - Make `igraph_integer_t` configurable, but support only a few reasonable choices (i.e. 32 or 64-bit signed integer).
 - There should be a _default_ configuration, which is to match the size of `size_t`. igraph should work in this mode, unless the user explicitly compiles it with a non-default config. Linux distros will include only the default configuration in their packages.
 - If some software that depends on C/igraph uses the non-default configuration, it should either link to C/igraph statically, or it should ship its own private copy. R/igraph already links to C/igraph statically, *even in Debian*, so nothing would need to change.

It seems to me that this would be a win-win: (1) we get the benefit of a larger `igraph_integer_t` (2) no extra maintenance burden on R/igraph (3) effectively no change for Linux distros and distro maintainers.

----

@gaborcsardi @ntamas, could you indicate if you'd both be happy with such a solution?

----

Some of the things I originally wrote were based on the incorrect assumption that distro maintainers would always want to enforce linking _dynamically_ to a single igraph library, even for the R interface. This is not the case.

**Comment by gaborcsardi** [comment]
> Explicit limits on graph and array sizes

These are defined by the data types, IDK if they are platform dependent or not, but my understanding is that this is a documentation issue, at least for the graph sizes?

> Move igraph_integer_t to 64-bit

This is not possible, because R does not have a 64 bit integer data type. Rewriting the R interface to get rid of the direct mapping is a lot of work, and it will also break a lot of existing R igraph code. If the new R igraph data type is just an external pointer to the C structure, like in Python, then this  will also mean a degraded user experience, because you won't be able to serialize igraph objects in R, meaning no `save()`, `load()`, saving the session, etc.

Second, I am not sure if this the right approach for igraph, because the graph algorithms we have are so very different. Some need larger integers, but others don't. I would just add large integer support to the ones that do, without breaking the (R) APIs. E.g. we do use arbitrarily large integers in some algorithms already, afair. So we could just do the same in the other algorithms that need it. (If people actually need this feature as well!)

Third, a very-very small fraction of users benefit from this, the people running into the integer overflow issues. Frankly, it seems way too much work for the benefit.

So what I propose is:

* Leave `igraph_integer_t` as it is.
* For every algorithm, use the integer types that are appropriate. It is not very important what types an algorithm uses internally, only the API is important.
* Write a (short) guide to tell you how to choose the right integer type.
* Identify the algorithms that need larger integers because users run into issues. Address each of the individually. This might mean switching to arbitrarily large integers (optionally?), or switching to 64 bit integers internally. Or doing nothing if the use case is marginal, and it is too much work.

**Comment by gaborcsardi** [comment]
Obviously, I am not super involved in the development any more, so you can make a different decision, this is not a dictatorship, for better or worse. :)

**Comment by iosonofabio** [comment]
> Obviously, I am not super involved in the development any more, so you can make a different decision, this is not a dictatorship, for better or worse. :)

Best post of the week, it really cheers me up 👍 

I've googled a bit. Python began moving to int64 19 years ago and the problem is definitely gone with Python 3 which will be the only supported version soon.

R hasn't started yet, ints are 32bit. If the community is as progressive as Python's, they'll be fine around year 2040: right before the rising ocean waters take over New York City.

Jokes apart, that makes `igraph_integer_t` untouchable in terms of API. Instead, let's switch to long int in most internal places, which will be 64 bits on most platforms. That should also clean up a lot of casting that @ntamas hates :-P

I'm not sure I get @gaborcsardi 's suggestion of approaching it case by case and return the right one type for each. I mean, if an R user wants a graph without edges but with 7 billion vertices (e.g. "people at risk for COVID"), what are we supposed to do?

**Comment by gaborcsardi** [comment]
> I'm not sure I get @gaborcsardi 's suggestion of approaching it case by case and return the right one type for each. I mean, if an R user wants a graph without edges but with 7 billion vertices (e.g. "people at risk for COVID"), what are we supposed to do?

Fail? :) Honestly, I don't think you can do a lot of useful things with a graph like that. I think it is perfectly fine to target graphs whose vertex and edge ids fit into a 32 bit integer.

What I mean by this, is that even for smaller graphs, an algorithm might need larger integer types internally. E.g. betweenness, closeness, etc. may easily overflow the 32 bit integer.

**Comment by szhorvat** [comment]
> > Move igraph_integer_t to 64-bit
> 
> This is not possible, because R does not have a 64 bit integer data type.

However, if we make the size of `igraph_integer_t` configurable, then we're fine, right? The R interface will continue using 32-bit, anyone else is free to use 64-bit if they wish.

I don't see any arguments against making this configurable so far (except not wanting to have two libraries installed at the same time—but that is *not* necessary as I outlines in the proposal)

I maintain that making it configurable will not be a lot of work, and will not bring any noticeable maintenance overhead.

----

> > Explicit limits on graph and array sizes
> 
> These are defined by the data types, IDK if they are platform dependent or not, but my understanding is that this is a documentation issue, at least for the graph sizes?

It's not merely a documentation issue.

If sizes (array lengths, vertex counts, edge counts, etc.) become too large, parts of the code will start failing gradually. Different parts of the code fail at different sizes (in the original post above I described several failure points). There are currently no practical guidelines for a user to see what is a maximum reasonable size.

If the failure happened consistently at a predictable size, and would always trigger an error message ("out of memory" or similar), then there would be no problem. But currently this is not what happens.

Therefore, I would like to have explicit size limits introduced and enforced. With this approach, when the limits are exceeded, igraph can fail with a clear and understandable error.

----

> because the graph algorithms we have are so very different. Some need larger integers, but others don't.

The main issue is not with the needs of internal algorithms, but limits on representable graph sizes. I'd like to resolve two problems:

 1. Optionally allow bigger graphs (when `igraph_integer_t` is configured to be 64-bit)
 2. (A) Make sure that current uses of integer types are consistent; and (B) Make sure that it is easy to check that limits are no exceeded when implementing algorithms.  

1. would be solved by moving to 64-bit (while retaining a 32-bit compilation mode) and 2. would be solved by explicitly enforced limits

----

> Instead, let's switch to long int in most internal places, which will be 64 bits on most platforms.

I would like to use types of a predictable size instead of raw C types whose size differs between platforms. There are ways to do this, e.g. `int64_t` or platform-specific configuration. Many (most?) libraries these days use predictable-size types.

----

As I said above, I do not see any good arguments against making `igraph_integer_t` configurable.  Let me know if I missed any.

There would be no change for the R interface (we can set it up so that defining `USING_R` already chooses 32-bit size).

**Comment by szhorvat** [comment]
To elaborate more, it seems to me that there are three options for the 32-bit/64-bit decision:

 - **Move to 64-bit:** This would require major changes to R/igraph, which we don't have the resources to do. It would break backwards compatibility of R/igraph (serialized igraph object). The disadvantages seem to far outweigh the benefits.
 - **Stay on 32-bit,** i.e. maintain the status quo. No cost, but also nothing gained. Over time, igraph will slowly lose competitiveness.
 - **Make the type configurable:** Has the advantages of both approaches, but the disadvantages of neither.

----

As for **making limits explicit**, as far as I could tell there was no serious objection.

**Comment by gaborcsardi** [comment]
> Stay on 32-bit, i.e. maintain the status quo. No cost, but also nothing gained. Over time, igraph will slowly lose competitiveness.

Do you really think that there are a lot of people that need to be able to handle graphs with more than 2 billion vertices/edges? What do you do with a graph of that size? You can basically only run linear algorithms, no?

But why don't we make a survey and ask people if they want the igraph team to
- fix existing bugs,
- add new algorithms,
- make the API more consistent,
- make igraph faster,
- improve documentation, or 
- add support for graphs with more than 2 billion vertices/edges.

**Comment by gaborcsardi** [comment]
Re making the integer size configurable, of course that is fine with me (but any solution you choose is). 

Personally, I do think that it is 
1. not a little amount of work, 
2. it is bloat that you'll need to support forever, 
3. people don't actually need larger than 2 billion node graphs, 
4. if it is configured to be 64 bit, then _all_ graphs need twice as much memory, and _all_ algorithms will be slower.

**Comment by szhorvat** [comment]
>     3\. people don't actually need larger than 2 billion node graphs,

We must think about edge-counts too, not just vertex counts. Edge counts are usually larger.

On 32-bit, the limit is actually closer to 200 million (see my estimation in the original post). But this is mostly due to address space limitations on 32-bit. If we introduce explicit limits, we could make them larger on 64-bit machine even if we stick to 32-bit indexing only. However, considering how common "edge lists" of the format returned by `igraph_get_edgelist` are, the practical limit will still be half of the max index, i.e. ~1 billion.

I quickly looked through the graphs on https://snap.stanford.edu/data/ and there is one with ~1.8 billion edges ("Friendster network").

I assumed that the dataset mentioned in #1435 was also a real-world one, but of course it could be a generated one ...

> You can basically only run linear algorithms, no?

Yes, but those are important for real-world network analysis. Example use case: load a large graph and find a representative subset/subgraph, on which it is feasible to run slow algorithms (e.g. with random-walk or with neighbourhood search).

>     4\. if it is configured to be 64 bit, then _all_ graphs need twice as much memory, and _all_ algorithms will be slower.

It won't be worse than the current version. Right now `igraph_t` has vectors of `double`s. A `double` takes up 64 bits.

We'd only gain some efficiency once transitioning to integers.

**Comment by szhorvat** [comment]
> * not a little amount of work,
> 
> * it is bloat that you'll need to support forever,

Perhaps I should show what I have in mind in a separate branch. Then we can discuss it in more concrete terms.

**Comment by gaborcsardi** [comment]
> On 32-bit, the limit is actually closer to 200 million (see my estimation in the original post).

In here? https://github.com/igraph/igraph/issues/1450#issue-666045702 Where?

**Comment by gaborcsardi** [comment]
> I wouldn't be more bloat than supporting 32-bit integers on 32-bit platforms and 64-bit integers on 64-bit platforms. 

32 bit platforms are practically dead. I honestly don't care if we don't support them at all. It is hard to maintain them as well, e.g. is there a 32 bit CI?

**Comment by vtraag** [comment]
Great discussion! I'm just now reading up on all of this. I actually didn't realize `igraph_integer_t` was still 32-bits (also on 64-bits machines).

I agree with @szhorvat that supporting graphs with more than about 1 billion edges should be feasible. Right now, this is not possible, even if you have 64-bits machines. The limitation of 1 billion edges is really not *that* uncommon, and will probably become increasingly more common.  At my institute, we also work with large citation networks that also have increasingly more often more than 1 billion edges, and we are also hitting 32-bits limitations.

It would be strange to pretend that `igraph` is a library that is able to work efficiently with large graphs, yet cannot handle graphs that would otherwise easily fit into available memory (on 64-bits machines, with enough memory of course). Clearly, most graphs will be smaller than 1 billion edges, and most users will not hit that limit. But if `igraph` would not support it, those users would need to switch to other libraries, and they may influence other people along with it.

In order to support the (rather surprising) lack of 64-bits integers in R, I think the proposal by @shorvzat is quite reasonable. It means that it is possible to support larger graphs (i.e. beyond the 1 billion edges), but that still allows R to only work with 32-bits integers.

> Perhaps I should show what I have in mind in a separate branch.

That would be a good idea! If I can help out with this, let me know.

---

One other problem might be that `igraph_t` uses doubles internally (`igraph_vector_t`), which only represents integers correctly up to 32-bits as well, if I understand correctly. That means that the `oi` and the `ii` vectors may not correctly index the `from` and `to` vectors, and similarly `os` and `is` may incorrectly index `oi` and `ii`. Hence, we would still not be able to support more than 1 billion edges correctly if we just make `igraph_integer_t` 64 bits, we would also need to switch the `igraph_t` type to 64 bits integers. If my understanding of integer support in doubles is correct, this would indeed be a much larger overhaul of the code.

I have some other suggestions for changing the graph type anyway, to support a number of operations more efficiently (see our [discussion](https://igraph.discourse.group/t/igraph-edge-and-from-to-swapping/310/8)) in the forum). I will post something on this shortly. I would like to make sure that for a 1.0 version we have a graph type that we are satisfied with and which we would like to support in the long term.

**Comment by vtraag** [comment]
> One other problem might be that `igraph_t` uses doubles internally (`igraph_vector_t`), which only represents integers correctly up to 32-bits as well, if I understand correctly.

I think this may actually be up to 52 bits, which is still reasonable, although it of course induced another edge limitation `IGRAPH_MAX_ECOUNT` than originally proposed.

**Comment by ntamas** [comment]
Okay, back from holidays. I have read and carefully considered everything, and I vote for making the size of the integer type configurable at compilation time, and make 64-bit the default. At the same time, let's alter the R interface and add a compile-time switch to keep on using 32-bits.

**Comment by szhorvat** [comment]
it sounds like we are converging towards a solution. We should finish up the transition to cmake before taking this on.

**Comment by ntamas** [comment]
Related issue: https://github.com/igraph/python-igraph/issues/338

It seems like people are indeed sometimes hitting the edge count limit.

**Comment by szhorvat** [comment]
Closing this, and opening a new issue for checking for overflow where necessary. If there's anything else to do for this issue that I missed, please reopen (or better: open a separate, more specific issue).
